### PR TITLE
fix: Fixes Cookie SameSite config validation

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change

(A few sentences about this PR)

## Related issues

-  https://github.com/supertokens/supertokens-golang/issues/141

## Test Plan

Added unit tests.

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here highlighting the necessary changes)

## Checklist for important updates

-   [x] Changelog has been updated
-   [ ] ~~`coreDriverInterfaceSupported.json` file has been updated (if needed)~~
    -   ~~Along with the associated array in `supertokens/constants.go`~~
-   [ ] ~~`frontendDriverInterfaceSupported.json` file has been updated (if needed)~~
-   [ ] Changes to the version if needed
    -   In `supertokens/constants.go > version variable`
-   [x] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.

### CODE REVIEW

File: recipe/session/config_test.go [hunk 1]
```diff
@@ -1252,3 +1254,66 @@ func TestSuperTokensInitWithDefaultAPIGateWayPathandCustomAPIBasePath(t *testing
...
+		AppInfo: supertokens.AppInfo{
+			AppName:       "SuperTokens",
+			WebsiteDomain: "http://supertokens.io",
+			APIDomain:     "http://supertokensapi.io",
+		},
+		RecipeList: []supertokens.Recipe{
+			Init(&sessmodels.TypeInput{}),
+		},
+	}
+
```
Comment (rishabhpoddar) [inline recipe/session/config_test.go:1314]: Add the following tests if they don't already exist:
- When using localhost or IP for both apiDomain and websiteDomain, adding sameSite none should be allowed
- When apiDomain and websiteDomain are both an IP or localhost with https, then none should be allowed
- When apiDomain and websiteDomain are both non localhost and non IP, but with https, then none should be allowed.
- When apiDomain is non localhost and non IP (with https), websiteDomain is http://localhost:3000, then none should be allowed.

Please make sure all the possible inputs are covered - this is a very important config.

File: recipe/session/config_test.go [hunk 2]
```diff
@@ -605,7 +606,8 @@ func TestSuperTokensWithAntiCSRFNone(t *testing.T) {
 		RecipeList: []supertokens.Recipe{
 			Init(
 				&sessmodels.TypeInput{
-					AntiCsrf: &customAntiCsrfVal,
+					AntiCsrf:     &customAntiCsrfVal,
+					CookieSecure: &True,
```
Comment (rishabhpoddar) [inline recipe/session/config_test.go:610]: why is this modified?

File: recipe/session/session_test.go [hunk 3]
```diff
@@ -657,7 +658,8 @@ func TestWhenAntiCsrfIsDisabledFromSTcoreNotHavingThatInInputToVerifySessionIsFi
 		},
 		RecipeList: []supertokens.Recipe{
 			Init(&sessmodels.TypeInput{
-				AntiCsrf: &customAntiCsrfVal,
+				AntiCsrf:     &customAntiCsrfVal,
+				CookieSecure: &True,
```
Comment (rishabhpoddar) [inline recipe/session/session_test.go:662]: why is this modified?

File: recipe/session/utils.go [hunk 4]
```diff
@@ -146,9 +146,8 @@ func validateAndNormaliseUserInput(appInfo supertokens.NormalisedAppinfo, config
 	}
 
 	if cookieSameSite == cookieSameSite_NONE &&
-		!cookieSecure &&
-		!(topLevelAPIDomain == "localhost" || IsAnIPAPIDomain) &&
-		!(topLevelWebsiteDomain == "localhost" || IsAnIPWebsiteDomain) {
+		!cookieSecure && (!(topLevelAPIDomain == "localhost" || IsAnIPAPIDomain) ||
+		!(topLevelWebsiteDomain == "localhost" || IsAnIPWebsiteDomain)) {
```
Comment (rishabhpoddar) [inline recipe/session/utils.go:150]:
```suggestion
		!cookieSecure && !((topLevelAPIDomain == "localhost" || IsAnIPAPIDomain) &&
		(topLevelWebsiteDomain == "localhost" || IsAnIPWebsiteDomain)) {
```

Comment (rishabhpoddar) [reply recipe/session/utils.go:150]: I find the above change easier to read, and also add comments explaining this line in detail

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: recipe/session/config_test.go (hunk 2 above)
Comment (sattvikc) [reply recipe/session/config_test.go:610]: Test was failing.

File: recipe/session/session_test.go (hunk 3 above)
Comment (sattvikc) [reply recipe/session/session_test.go:662]: Test was failing.

Comment (rishabhpoddar) [reply recipe/session/session_test.go:662]: why?

Comment (sattvikc) [reply recipe/session/session_test.go:662]: image

This combination is not allowed. Do you prefer this over secure cookie?

image

**Review by rishabhpoddar — APPROVED** [review]
